### ISSUE TITLE
Feature Request: Unlock the full potential of the concat script demuxer.

### ISSUE THREAD

**Issue by baronsmedley** [Nevcairiel/LAVFilters#299]
Recently, I worked on a little project. After doing some research, I realized that it would be possible to play video directly off of a Windows named pipe with the use of the ffmpeg concat script demuxer. The concat script demuxer has the ability to play video accessed by any protocol available to ffmpeg directly from a file. And yes, this includes Windows named pipes. Of course, you can also play video accessed by all the other protocols available to ffmpeg. This includes HTTP, TCP, UDP ect. With this feature fully enabled, the lav filters become very powerful. It enables directshow players to play from these protocols even if they don't have the inherent ability to do so.

 I've made some simple alterations to ffmpeg's source code using the lav filters version of the code. After compiling the libraries, I simply copied my version of the ffmpeg libraries into the lav filters x86 folder. This proved to be a lot easier than I thought it would be. After some testing, I found that I had indeed fully unlocked the concat script demuxer. All my tests worked perfectly.

 I'll spell out the changes to the code that I made, however, you may wish to do a little bit more research. One of the things that I did was enable a bunch of protocols which by default are disabled in ffmpeg. The method that I chose to do this may be too general and simplistic for you to choose. Some time ago, ffmpeg instituted a protocol whitelist. Most of the protocols are blacklisted and if you want to access them you have to place them on a whitelist on the command line. I don't know why ffmpeg Instituted this feature: it didn't used to have it. I can only assume that this was done to address security issues. But, since ffmpeg only really deals with audio video, I can't see how ffmpeg can be leveraged to compromise a computer. But then again, I'm not a hacker. After peering through the code, I found that any file that ffmpeg accesses inherits its whitelist from file.c located under libavformat. I simply change the line (.default_whitelist   = "file,crypto") to (.default_whitelist  = "file,crypto,http,https,tls,rtp,tcp,udp,rtmp"). Although this worked, this seems to universally enable these protocols. You may choose to have the default whitelist redeclared in concatdec.c in order to enable these protocols specifically for the contact script demuxer. I don't really know if that would be better or even possible. I'm no expert in C, however, I can brush up and program in just about any language when I need to.

 I only had to make two more changes in ffmpeg. Both of them are very straightforward and simple one liners. The concat script demuxer has a option called safe mode which is enabled by default. With this option enabled, the concat script demuxer can only access files with relative file paths. This option must be disabled. In concatdec.c under libavformat simply change (OFFSET(safe), AV_OPT_TYPE_BOOL, {.i64 = 1}, -1, 1, DEC },) to (OFFSET(safe), AV_OPT_TYPE_BOOL, {.i64 = 0}, -1, 1, DEC },).

 The third change that I made is not within the ffmpeg source code itself but rather a change to your compilation script. I use your build_ffmpeg.sh script to build ffmpeg. Within this script I deleted the line --disable-bsfs. The concat script demuxers intended purpose is to splice video files together. In order to do this, it presents every file it accesses as a streamable source (no seeking allowed). Obviously if a file is not streamable like say an mp4, the demuxer requires a bit stream filter to convert it. I'm pretty sure the only one that it needs is h264_mp4toannexb. This is actually stated in ffmpegs documentation -- "Currently, the only conversion is adding the h264_mp4toannexb bitstream filter to H.264 streams in MP4 format." I got lazy and just enabled all of them. You may want to do some research to ensure that the h264_mp4toannexb bitstream filter is the only one required.

 I highly recommend that you enable this feature. I'm now using it extensively in combination with SageTV. I wrote a Windows batch script to present a live stream from an h.264 encoder over a Windows named pipe on demand. I muxed the video and audio into a streamable Avi using ffmpeg. ffmpeg is started when the pipe is accessed and quits when the pipe is disconnected. The script then loops and awaits another connection. The script also sends events to eventghost. This allows me to switch my remote control over from controlling SageTV to controlling my stb directly. When I exit the viewer, remote control is reverted back to controlling SageTV. The encoder that I'm using is very low latency and I can control my stb comfortably. This is now how I watch live TV  when I am not recording. It's really pretty fantastic. No file caching, no latency, and I am able to use all of the features that my stb has to offer. I'm also able to use SageTV normally the way it was intended.

**Comment by Nevcairiel** [comment]
The protocol whitelist exists for a reason, because a file causing any software to do web requests is a huge security hole waiting to happen, nevermind the second change to allow any file on the system to be accessed without any regards for security, which is also why I have no plans to make such changes.

**Comment by baronsmedley** [comment]
As I wrote in my request I had surmised this as well, however I've been using the lav filters to play m3u8 files  for quite some time now.  It appears that you enabled this feature which is in direct contradiction to your response. Why would an FFcontact file be any more dangerous than an m3u8 file.  I'm sorry I don't get it.  As I stated I'm not a hacker so I don't really know but it appears that you've already done this.
